res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,774 @@
+
+        engine.setValue(group, "pitch", newValue);
+        engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandplay", 1);
+    } else {
+        // decrease the number of active Pads, this should minimize unwanted stops
+        this.keyboardModeRefCount[channel] -= 1;
+        if (this.keyboardModeRefCount[channel] <= 0) {
+            engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandstop", 1);
+            engine.setValue(group, "pitch", 0.0); // reset pitch
+            this.keyboardModeRefCount[channel] = 0; // reset refcount to 0
+        }
+    }
+};
+
+PioneerDDJ400.keyshiftModePad = function(_channel, control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+    engine.setValue(group, "pitch", this.halftoneToPadMap[control & 0xf]);
+};
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+
+    var playing = engine.getValue(group, "play");
+    // when playing stop and return to start/cue point
+    if (playing > 0) {
+        engine.setValue(group, "cue_gotoandstop", 1);
+    } else { // load selected track
+        engine.setValue(group, "LoadSelectedTrack", 1);
+    }
+};
+
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value) {
+    "use strict";
+    this.shiftState[channel] = value;
+};
+
+PioneerDDJ400.waveFormRotate = function(_channel, _control, value) {
+    "use strict";
+    // select the Waveform to zoom left shift = deck1, right shift = deck2
+    var deckNum = this.shiftState[0] > 0 ? 1 : 2;
+    var oldVal = engine.getValue("[Channel"+deckNum+"]", "waveform_zoom");
+    var newVal = oldVal + (value > 0x64 ? 1 : -1);
+
+    engine.setValue("[Channel"+deckNum+"]", "waveform_zoom", newVal);
+};
+
+PioneerDDJ400.loopin4beatPressed = function(channel, _control, value, _status, group) {
+    "use strict";
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    this.loopin4beat[channel] = (value > 0);
+
+    if (loopEnabled === 0 && value > 0) {
+        engine.setValue(group, "loop_in", 1);
+    }
+};
+
+PioneerDDJ400.loopin4beatPressedLong = function(_channel, _control, value, _status, group) {
+    // problematic - loop gets set to the playback position where the 'long press' was recognized
+    // and not to the point at which the button was initially pressed
+    // as a result, the loop is not set where one would expect
+    "use strict";
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    if (!loopEnabled && value > 0) {
+        engine.setValue(group, "beatloop_4_activate", 1);
+    }
+};
+
+PioneerDDJ400.loopoutPressed = function(channel, _control, value, _status, group) {
+    "use strict";
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    this.loopout[channel] = (value > 0);
+
+    if (loopEnabled === 0 && value > 0) {
```

Comment:
So after some experimentation, I think I can see why this was originally implemented in this manner.

As written (and consistent with how the DDJ-400 documentation is written), when a loop is active, one can press and hold the Loop In button and then use the jogwheel to "micro adjust" the loop start point.

If I remove this condition, and instead just call engine.setValue(group, "loop_in", value), then pressing the Loop In button immediately moves the loop start point to the current position (respecting quantization if set).  One then no longer has the ability to "micro adjust" the loop start point using the jogwheel.  This is consistent with the loop_in documentation at https://www.mixxx.org/wiki/doku.php/mixxxcontrols

By "micro adjust the loop start point" what I mean is that the existing loop start point is shifted by a small amount in the direction indicated by the jogwheel turn direction, as controlled by the loopAdjustMultiply variable.

So, in conclusion, the current implementation is consistent with the DDJ-400 documentation.  Making the above changes would make the behavior consistent with the Mixxx control documentation.  However doing so would eliminate functionality that's described in the DDJ-400 docs.

My personal inclination would be to retain the "micro adjust" functionality and document it on the DDJ-400 wiki page.

Thoughts?

Reply:
I understand how that is more useful than the standard Mixxx behavior, but I don't think a controller script is a good place to implement this. Instead the behavior of `loop_in` and `loop_out` should be changed in Mixxx. Regardless, the script never resets `loop_in` or `loop_out` to 0, which is unexpected behavior not guaranteed to do what you want.

Reply:
Ok, I'll cut this too (while making a note in the comment section as above). These advanced loop features aren't really that profound if we just add a simple quantize toggle (I map one to `shift + beat sync`). Even on high end club gear, I wouldn't loop live without quantize set...

Reply:
What needs to happen here I think is:
```js
if (loopEnabled) {
     // ... DDJ400 jog adjust stuff
}
else {
    engine.setValue(group, "loop_out", value); // ensures loop_out gets 0 when released
}
```

Reply:
I'm not sure about that. That would only work if you can guarantee that `loop_enabled` is 0 when the button is released.

Reply:
> I'm not sure about that. That would only work if you can guarantee that loop_enabled is 0 when the button is released.
Apologies---I deleted the comment this was referring to before seeing your reply.

I've ended up going back to a simple shift mode implementation of the advanced feature. It will guarantee `loop_enabled` to be called with 0 as required.